server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -680,13 +680,28 @@ public Map<String, Object> readMap() throws IOException {
         return (Map<String, Object>) readGenericValue();
     }
 
+    public <K, V> Map<K, V> readImmutableMap(Writeable.Reader<K> keyReader, Writeable.Reader<V> valueReader) throws IOException {
```

Comment:
Can we add a comment here, making it clear that `valueReader` must never return `null`? I see we're using the immutable open map with optional strings for the values, that will throw NPEs with `Map.of(` and can't be ported directly. Might be nice to make that clear in the docs?